### PR TITLE
add dynamic sdk support for topology extensions

### PR DESCRIPTION
**Pull request by glekner**

### CODE REVIEW

File: frontend/packages/topology/src/components/graph-view/Topology.tsx [hunk 1]
```diff
@@ -138,10 +143,14 @@ const Topology: React.FC<TopologyProps &
   const componentFactoryExtensions = useExtensions<TopologyComponentFactory>(
     isTopologyComponentFactory,
   );
+  const [dynamicComponentFactoryExtensions, isFactoryComponentsResolved] = useResolvedExtensions<
+    DynamicTopologyComponentFactory
+  >(isDynamicTopologyComponentFactory);
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/graph-view/Topology.tsx:148]: Can we move away from `useExtensions` for the old extension and make use of `useResolvedExtensions`?
Then combine the results two variables: one for the resolved state and one for an array of factories.

This way it'll be two code paths for setup but consumption will be simple.

File: frontend/packages/topology/src/components/graph-view/Topology.tsx [hunk 2]
```diff
@@ -230,12 +239,18 @@ const Topology: React.FC<TopologyProps &
...
+            ...dynamicComponentFactoryExtensions.map((f) => f.properties.getFactory),
+          ]);
+        })
+        .catch(() => {});
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [componentFactoriesPromises, isFactoryComponentsResolved]);
 
   React.useEffect(() => {
     if (componentFactoriesPromises.length && !componentFactories.length) {
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/graph-view/Topology.tsx:256]: Since you've separated the static and dynamic extensions into different variables, this check is no longer valid as it only applies to one.

See other comment about using `useResolvedExtensions` for both static and dynamic plugins.

File: frontend/packages/topology/src/data-transforms/DataModelProvider.tsx [hunk 3]
```diff
@@ -18,13 +23,23 @@ const DataModelProvider: React.FC<DataModelProviderProps> = ({ namespace, childr
   }, [namespace]);
 
   const modelFactories = useExtensions<TopologyDataModelFactory>(isTopologyDataModelFactory);
+  const [dynamicModelFactories, isModelResolved] = useResolvedExtensions<
+    DynamicTopologyDataModelFactory
+  >(isDynamicTopologyDataModelFactory);
```
Comment (christianvogt) [inline frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:28]: Can both paths use `useResolvedExtensions` so that we don't have one path doing manual resolution.

File: frontend/packages/topology/src/components/dropdowns/ApplicationDropdown.tsx [hunk 4]
```diff
@@ -40,10 +45,13 @@ interface ApplicationDropdownProps {
 const ApplicationDropdown: React.FC<ApplicationDropdownProps> = ({ namespace, ...props }) => {
   const { t } = useTranslation();
   const modelFactories = useExtensions<TopologyDataModelFactory>(isTopologyDataModelFactory);
+  const [dynamicModelFactories] = useResolvedExtensions<DynamicTopologyDataModelFactory>(
+    isDynamicTopologyDataModelFactory,
+  );
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/dropdowns/ApplicationDropdown.tsx:50]: Can we update the static extension to make use of `CodeRef` then simplified the resolution with `useResolvedExtensions` 

We probably don't want to render the firehose component until all extensions have been resolved.

**Comment by glekner** [comment]
@christianvogt modified extensions, now consuming all with `useResolvedExtensions`

File: frontend/packages/helm-plugin/src/topology/helmFilters.ts [hunk 5]
```diff
@@ -35,3 +36,14 @@ export const applyHelmDisplayOptions = (model: Model, filters: DisplayFilters):
 };
 
 export const applyDisplayOptions = () => applyHelmDisplayOptions;
+
+export const getHelmWatchedResources = (namespace: string): WatchK8sResources<any> => {
```
Comment (christianvogt) [inline frontend/packages/helm-plugin/src/topology/helmFilters.ts:40]: Why was this moved to a filters module?

File: frontend/packages/rhoas-plugin/src/topology/createConnector.ts [hunk 6]
```diff
@@ -19,3 +22,16 @@ export const getCreateConnector = (createHints: string[], source: Node, target:
   }
   return null;
 };
+
+export const getRhoasWatchedResources = (
```
Comment (christianvogt) [inline frontend/packages/rhoas-plugin/src/topology/createConnector.ts:26]: Moving to `createConnector` doesn't seem like a logical home.

File: frontend/packages/topology/src/operators/operatorFilters.ts [hunk 7]
```diff
@@ -35,3 +39,25 @@ export const applyOperatorDisplayOptions = (model: Model, filters: DisplayFilter
 };
 
 export const applyDisplayOptions = () => applyOperatorDisplayOptions;
+
+export const getOperatorWatchedResources = (namespace: string): WatchK8sResources<any> => {
```
Comment (christianvogt) [inline frontend/packages/topology/src/operators/operatorFilters.ts:43]: Same comment about location.

**Review by christianvogt — COMMENTED** [review]
Operator backed services aren't showing up correctly. Use the dev catalog
Before:
image

After:
image

There's also flashing of the old value when switching from one namespace to another.

**Comment by christianvogt** [comment]
Display filters are missing:

Before:
image

After:
image

**Comment by christianvogt** [comment]
@glekner the latest changes fixed the missing visuals.

One outstanding issue remains. Flashing of the old value when switching from one namespace to another.
In the following gif, the visual of the operator backed service disappears when we change namespaces (expected), then then reappears (unexpected) before the proper topology for the new namespace is displayed.

topology-flash

**Comment by glekner** [comment]
hey again,
the flashing is fixed by reverting `ModelFactory` to use `useExtensions`.
since you can’t pass a function without a CodeRef for the dynamic extensions, instead we are passing the resources that need to be watched, and we build a function for them in the consuming of the extensions. please review @christianvogt

**Comment by glekner** [comment]
/test e2e-gcp-console



<br class="Apple-interchange-newline">

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts [hunk 8]
```diff
@@ -88,7 +88,7 @@ export const isTopologyCreateConnector = (e: Extension): e is TopologyCreateConn
 export const isTopologyDataModelFactory = (e: Extension): e is TopologyDataModelFactory =>
   e.type === 'console.topology/data/factory';
 
-export const isTopologyDisplayFilters = (e: Extension): e is TopologyDisplayFilters =>
+export const isTopologyDisplayFilter = (e: Extension): e is TopologyDisplayFilters =>
```
[the remark this answers is not in the document]
Comment (glekner) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts:91]: to match the static extension

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts [hunk 9]
```diff
@@ -39,7 +39,7 @@ export type TopologyDataModelFactory = ExtensionDeclaration<
     /** Priority for the factory */
     priority: number;
     /** Resources to be fetched from useK8sWatchResources hook. */
-    resources?: CodeRef<(namespace: string) => WatchK8sResources<any>>;
+    resources?: WatchK8sResources<any>;
```
[the remark this answers is not in the document]
Comment (glekner) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts:42]: since you cant set a function without a CodeRef, i'm building the namespaced routes here
https://github.com/openshift/console/pull/9055/files#diff-281f5462e39247e7a7894ee5ad688563cf18b46d97e7707da0b660b79c236f41R17

File: frontend/packages/topology/src/components/dropdowns/ApplicationDropdown.tsx [hunk 10]
```diff
@@ -56,7 +69,9 @@ const ApplicationDropdown: React.FC<ApplicationDropdownProps> = ({ namespace, ..
       ...watchedBaseResources[key],
       prop: key,
     }));
-  }, [modelFactories, namespace]);
+
+    return [];
```
[the remark this answers is not in the document]
Comment (glekner) [reply frontend/packages/topology/src/components/dropdowns/ApplicationDropdown.tsx]: thanks, removed

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (rohitkrai03) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts:42]: So why not make it a coderef?

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rohitkrai03) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts:91]: Okay, I guess it was a typo in the static extensions. But we can keep it like this for now.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology.ts:42]: I talked to @glekner  about this too. I don't see an issue with this change unless the extension wants to hard code a namespace. In which case we can make sure to not set the namespace if one is already provided.

I don't think we have such a use case at this time that will be negatively affected by this.

Note that we still need to review the topology extensions for their exposed API and at that time may change things.

File: frontend/packages/topology/src/data-transforms/DataModelProvider.tsx [hunk 11]
```diff
@@ -10,6 +14,23 @@ interface DataModelProviderProps {
...
+) =>
+  factories.map(({ properties, ...ext }) => ({
+    ...ext,
+    properties: {
+      ...properties,
+      resources: (ns: string) =>
+        Object.assign(
+          {},
+          ...Object.entries(properties.resources).map(([k, v]) => ({
+            [k]: { ...v, namespace: ns },
```
Comment (christianvogt) [inline frontend/packages/topology/src/data-transforms/DataModelProvider.tsx:28]: We should probably accept the namespace provided before overriding? Not sure we have a use case for this yet but if an extension wanted to pull resources from a specific namespace we'd have to do:

```suggestion
            [k]: { namespace: ns, ...v },
```

**Review by rohitkrai03 — APPROVED** [review]
/lgtm
/approve
